yugabyte_dhstore.go

```diff
@@ -0,0 +1,159 @@
+package dhstore
+
+import (
+	"database/sql"
+	"fmt"
+
+	"github.com/lib/pq"
+	"github.com/multiformats/go-multihash"
+)
+
+type yugabyteConfig struct {
+	Host        string
+	Port        int
+	DBName      string
+	DBUser      string
+	DBPassword  string
+	SSLMode     string
+	SSLRootCert string
+}
+
+func NewYugabyteConfig() *yugabyteConfig {
+	return &yugabyteConfig{
+		Host:        "127.0.0.1",
+		Port:        5433,
+		DBName:      "yugabyte",
+		DBUser:      "yugabyte",
+		DBPassword:  "yugabyte",
+		SSLMode:     "disable",
+		SSLRootCert: "",
+	}
+}
+
+type yugabyteDHStore struct {
+	db *sql.DB
+}
+
+func NewYugabyteDHStore(c *yugabyteConfig) (DHStore, error) {
+	if c == nil {
+		c = NewYugabyteConfig()
+	}
+	psqlInfo := fmt.Sprintf("host=%s port=%d user=%s password=%s dbname=%s",
+		c.Host, c.Port, c.DBUser, c.DBPassword, c.DBName)
+
+	if c.SSLMode != "" {
+		psqlInfo += fmt.Sprintf(" sslmode=%s", c.SSLMode)
+
+		if c.SSLRootCert != "" {
+			psqlInfo += fmt.Sprintf(" sslrootcert=%s", c.SSLRootCert)
+		}
+	}
+
+	db, err := sql.Open("postgres", psqlInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	ydb := &yugabyteDHStore{
+		db: db,
+	}
+
+	ydb.createDatabase()
+	return ydb, nil
+
+}
+
+func (y *yugabyteDHStore) createDatabase() error {
+	stmt := `CREATE TABLE IF NOT EXISTS Multihash (
+                        mh BYTEA PRIMARY KEY,
+                        evks BYTEA[])`
+
+	_, err := y.db.Exec(stmt)
+	if err != nil {
+		return err
+	}
+
+	stmt = `CREATE TABLE IF NOT EXISTS  Metadata (
+                        hvk BYTEA PRIMARY KEY,
+                        emd BYTEA)`
+
+	_, err = y.db.Exec(stmt)
+	if err != nil {
+		return err.(*pq.Error)
+	}
+
+	return nil
+}
+
+func (y *yugabyteDHStore) MergeIndex(mh multihash.Multihash, evk EncryptedValueKey) error {
```

Comment:
We could certainly compile a multi-insert statement. However, under the hood it'd have to be balanced around the cluster, right? So regardless of how well we batch things up together, the db would have to decompose the batch and route right pieces to the right places under the hood.  

I'm open to any suggestions - please feel free to propose changes here.

Reply:
we should use batch insert in SQL engine i think.

Reply:
But this needs interface changes. Maybe we should do the interface changes first and move the "one at a time" insertion complexity into pebble instead of leaking it at interface level.

I can pick that up

Reply:
batch interface changes: https://github.com/ipni/dhstore/pull/49